subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -85,15 +102,21 @@ public DependentsSet getRelevantDependents(String className, IntSet constants) {
         if (!constants.isEmpty()) {
             return DependentsSet.dependencyToAll();
         }
-        Set<String> dependingOnAllOthers = annotationProcessingData.getGeneratedTypesDependingOnAllOthers();
-        if (deps.getDependentClasses().isEmpty() && dependingOnAllOthers.isEmpty()) {
+        Set<String> classesDependingOnAllOthers = annotationProcessingData.getGeneratedTypesDependingOnAllOthers();
+        Set<GeneratedResource> resourcesDependingOnAllOthers = annotationProcessingData.getGeneratedResourcesDependingOnAllOthers();
+        if (deps.getDependentClasses().isEmpty() && classesDependingOnAllOthers.isEmpty() && resourcesDependingOnAllOthers.isEmpty()) {
             return deps;
         }
-        Set<String> result = new HashSet<String>();
-        recurseDependents(new HashSet<String>(), result, deps.getDependentClasses());
-        recurseDependents(new HashSet<String>(), result, dependingOnAllOthers);
-        result.remove(className);
-        return DependentsSet.dependents(result);
+
+        // Recursively accumulate dependent classes and resources.  Dependent classes discovered can themselves be used to query
```

Comment:
Please don't use inline comments. Instead, if a class or method contains a complex algorithm, add a class or method level javadoc comment explaining the approach.